### PR TITLE
9533 query rewriter

### LINKED ISSUE #9533: Map instantiation fails when map is created from shared dataset
Working on #9516 I faced a trouble related to the query and the map instantiation. Follow theses steps to reproduce it:

- Share a dataset with write permission. Don't qualified the sql query (aka, include the username as part of the table name in the query).
- Using the user that it's shared with, create a map from this dataset.
- 💥  `Dashboard has some errors: Error: Map instantiation failed`.

This is happening because the queries present in the vizJSON are not qualified with the owner's username. 

screen shot 2016-08-24 at 11 50 16

This is hard to fix in the frontend (in the right way) because the map is initially instantiated in the deep-insight (and cartodb.js) side, from the data coming from the vizJSON. I tried to patch the vizJSON to fix the queries, and it solves the issue, but the solution smells. I feel the queries should come qualified in the vizJSON, in the same manner as `table_name` properties.

### PR DESCRIPTION
**Pull request by javitonino**
Qualifies table names in queries when creating a map from a shared dataset.

Also does a refactor trying to unify some of the query rewriting code. It is actually not needed with the latest version of this patch (I first thought that I could reuse the same query rewriting than in the importer for this, but qualifying != schema rewriting). I still kept it as I think it is a nice addition.

Fixes #9533

### CODE REVIEW

**Comment by javitonino** [comment]
Please CR @juanignaciosl

File: app/models/carto/analysis_node.rb [hunk 1]
```diff
@@ -55,6 +58,20 @@ def source_descendants
     children.map(&:source_descendants).flatten
   end
 
+  def fix_analysis_node_queries(old_username, new_user, renamed_tables)
```
Comment (juanignaciosl) [inline app/models/carto/analysis_node.rb]: Having two parameters for usernames and one for tables feels weird. Maybe splitting this method into `rename_analysis_node_queries_username` and `rename_analysis_node_queries_tables` would be more natural. Minor, though.

Comment (juanignaciosl) [inline app/models/carto/analysis_node.rb]: Meh, `rewrite_query` requires all three as well. Forget about that then, sorry 🙊

File: lib/carto/query_rewriter.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+module Carto
+  module QueryRewriter
+    def rewrite_query(query, old_username, new_user, renamed_tables)
+      new_query = query
```
Comment (juanignaciosl) [inline lib/carto/query_rewriter.rb]: Maybe it's out of scope and you prefer keeping this similar to previous version, but a `if .. elsif .. else` would be more readable.

File: app/models/map/copier.rb [hunk 3]
```diff
@@ -43,10 +43,12 @@ def copy_data_layers(origin_map, destination_map)
 
       attr_reader :map
 
-      def data_layer_copies_from(map)
-        map.carto_and_torque_layers.map { |layer|
-          layer.copy
-        }
+      def data_layer_copies_from(map, user)
```
Comment (juanignaciosl) [inline app/models/map/copier.rb]: Blind dog!

**Comment by juanignaciosl** [comment]
LGTM 👍  Feel free to ignore the only actual comment. :shipit: !

**Pull request merged**